WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Shipping Address Validation/ShippingLabelAddressFormViewModel.swift

```diff
@@ -75,45 +80,126 @@ final class ShippingLabelAddressFormViewModel: NSObject {
-            rows.removeAll(where: { $0 == .fieldError })
+        else if localErrors.contains(.city) {
+            if let index = rows.firstIndex(where: { $0 == .city }) {
+                rows.insert(.fieldError(.city), at: rows.index(after: index))
+            }
+        }
+        else if localErrors.contains(.postcode) {
+            if let index = rows.firstIndex(where: { $0 == .postcode }) {
+                rows.insert(.fieldError(.postcode), at: rows.index(after: index))
+            }
+        }
+        else if localErrors.contains(.state) {
+            if let index = rows.firstIndex(where: { $0 == .state }) {
+                rows.insert(.fieldError(.state), at: rows.index(after: index))
+            }
         }
-        return [Section(rows: rows)]
+        else if localErrors.contains(.country) {
+            if let index = rows.firstIndex(where: { $0 == .country }) {
+                rows.insert(.fieldError(.country), at: rows.index(after: index))
+            }
+        }
+        sections = [Section(rows: rows)]
     }
 }
 
-// MARK: - Remote API
+// MARK: - Local Validation
 extension ShippingLabelAddressFormViewModel {
-    func validateAddress(onSuccess: ((Bool, ShippingLabelAddressValidationError?) -> ())? = nil) {
+
+    enum ValidationError {
+        case name
+        case address
+        case city
+        case postcode
+        case state
+        case country
+    }
+
+    private func validateAddressLocally() -> [ValidationError] {
+        var errors: [ValidationError] = []
+
+        if let addressToBeValidated = address {
+            if addressToBeValidated.name.isEmpty {
+                errors.append(.name)
+            }
+            if addressToBeValidated.address1.isEmpty {
+                errors.append(.address)
+            }
+            if addressToBeValidated.city.isEmpty {
+                errors.append(.city)
+            }
+            if addressToBeValidated.postcode.isEmpty {
+                errors.append(.postcode)
+            }
+            if addressToBeValidated.state.isEmpty {
+                errors.append(.state)
+            }
+            if addressToBeValidated.country.isEmpty {
+                errors.append(.country)
+            }
+        }
+
+        return errors
+    }
+}
+
+// MARK: - Remote Validation API
+extension ShippingLabelAddressFormViewModel {
+
+    /// Validate the address locally and remotely. If `onlyLocally` is equal `true`, the validation will happens just locally.
+    ///
+    func validateAddress(onlyLocally: Bool, onSuccess: ((Bool, ShippingLabelAddressValidationError?) -> ())? = nil) {
+
+        addressValidationError = nil
+        if validateAddressLocally().isNotEmpty {
+            isAddressLocallyValidated = false
+            state.isLoading = false
+            onSuccess?(false, nil)
+            return
+        }
+        isAddressLocallyValidated = true
+
+        if onlyLocally {
+            onSuccess?(true, nil)
+            return
```

Comment:
nit: I think you can write this as 
```swift
return onSucess?(...)
```
Which looks a tad bit nicer!

Reply:
Unfortunately, it doesn't work, because in this case, the function expects a non-void return value 🤕